### PR TITLE
DropDownButton: selectedItem should be kept after dataSource reload when new dataSource includes selectedItemKey (T919804)

### PR DESCRIPTION
**Pull request by ksercs**

### CODE REVIEW

File: js/ui/drop_down_button.js [hunk 1]
```diff
@@ -518,9 +518,22 @@ const DropDownButton = Widget.inherit({
         }
     },
 
+    _getItemByKey() {
+        const key = this._getKey();
+        const selectedItemKey = this.option('selectedItemKey');
+        const items = this.option('items');
+        const filter = item => (key === 'this' ? item : item[key]) === selectedItemKey;
+
+        return items?.find(filter) || null;
```
Comment (DokaRus) [inline js/ui/drop_down_button.js]: IE11 [does not support](https://caniuse.com/#feat=array-find) it :(

Comment (ksercs) [reply js/ui/drop_down_button.js]: fixed

File: js/ui/drop_down_button.js [hunk 2]
```diff
@@ -518,9 +518,22 @@ const DropDownButton = Widget.inherit({
         }
     },
 
+    _getItemByKey() {
+        const key = this._getKey();
+        const selectedItemKey = this.option('selectedItemKey');
+        const items = this.option('items');
```
Comment (alexander-kotov-dx) [inline js/ui/drop_down_button.js]: Does it work for data from the 'dataSource' option?

File: js/ui/drop_down_button.js [hunk 3]
```diff
@@ -520,8 +523,19 @@ const DropDownButton = Widget.inherit({
 
     _updateItemCollection(optionName) {
         this._setWidgetOption('_list', [optionName]);
-        this._setListOption('selectedItemKeys', []);
-        this._loadSelectedItem().done(this._updateActionButton.bind(this));
+        this.getDataSource().store().byKey(this.option('selectedItemKey')).done(selectedItem => {
```
Comment (ksercs) [inline js/ui/drop_down_button.js]: This `byKey` call raises the `Data item cannot be found` error when there is no necessary item in the new dataSource. Is it okay?

File: js/ui/drop_down_button.js [hunk 4]
```diff
@@ -522,9 +528,24 @@ const DropDownButton = Widget.inherit({
     },
 
     _updateItemCollection(optionName) {
+        const selectedItemKey = this.option('selectedItemKey');
         this._setWidgetOption('_list', [optionName]);
-        this._setListOption('selectedItemKeys', []);
-        this._loadSelectedItem().done(this._updateActionButton.bind(this));
+
+        if(isDefined(selectedItemKey)) {
+            this.getDataSource().store().byKey(selectedItemKey).done(selectedItem => {
```
Comment (DokaRus) [inline js/ui/drop_down_button.js]:
```suggestion
            this.getDataSource()
                .store()
                .byKey(selectedItemKey)
                .done(selectedItem => {
```

**Review by DokaRus — DISMISSED** [review]

**Review by DokaRus — DISMISSED** [review]

**Review by DokaRus — DISMISSED** [review]

File: js/ui/drop_down_button.js [hunk 5]
```diff
@@ -568,12 +592,16 @@ const DropDownButton = Widget.inherit({
...
                 break;
             case 'items':
-                this._dataSource = null;
-                this._itemsToDataSource();
+                this._updateDataSource(this.option('items'));
                 this._updateItemCollection(name);
                 break;
             case 'dataSource':
-                this._initDataSource();
+                if(value instanceof Array) {
```
Comment (DokaRus) [inline js/ui/drop_down_button.js]: Why not use `Array.isArray`?

Comment (ksercs) [reply js/ui/drop_down_button.js]: fixed

**Review by DokaRus — APPROVED** [review]

**Review by alexander-kotov-dx — APPROVED** [review]

**Pull request merged**